Provide JARs with OSGi metadata

rotty3000 opened the pull request:
fixes  #195

Signed-off-by: Raymond Augé <raymond.auge@liferay.com>

## Overview

Provide JARs with OSGi metadata

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

sbrannen commented:
Tentatively slated for 5.6 M2 for _team discussion_

sbrannen commented:
I'd really like to see some integration tests in place that actually verify that the generated OSGi metadata is consumable.

@junit-team/junit-committers, thoughts?

rotty3000 commented:
Great feedback. I wanted to make the first pass small concise. I'm happy to make any adjustments and I don't see any issues making the adjustments you want. Finally I can surely add a verification of the metadata. :)

rotty3000 commented:
> Where is the Export-Package statement?
fixed
> API guardian must not be imported
fixed
> Kotlin must always be an optional import.
wonderful
> Why is org.opentest4j the only import with version information?
good question! working on it.

rotty3000 commented:
made `Bundle-Name` === `${project.description}`

sbrannen commented:
Did you forget to push the changes? Or... are you still working on the updates locally before pushing? (but.... absolutely no rush)

rotty3000 commented:
I've pushed all the above changes now. A major addition is only exporting packages marked with `@API` from all jars.

However there is an issue pending that I'm not sure why but for some reason bnd is not identifying the import versions for project dependencies. This is strange since I do see that bnd has the correct classpath and the jars are there. For example `junit-platform-engine` does not properly import `org.junit.platform.commons.*` with versions. I need to consult with @bjhargrave to figure out what might be the cause.

Also pending is verification of the metadata.

rotty3000 commented:
@sbrannen is it known that it's not possible to build any single module that has project dependencies? For instance when I try `./gradlew :junit-jupiter-engine:jar` it fails with
```
Execution failed for task ':junit-platform-commons:compileModule'.
> Extension with name 'sourceSets' does not exist. Currently registered extension names: [ext, eclipse, idea, defaultArtifacts, spotless]
```
it's not a huge deal, just makes debugging the build pretty hard if I can only execute the entire thing.

sbrannen commented:
> `./gradlew :junit-jupiter-engine:jar`

I just executed that, and it worked for me.

Maybe you did a `clean` first?

sbrannen commented:
Hmmm.... `./gradlew clean :junit-jupiter-engine:jar` works for me as well.

You executed that against `master` -- right?

sbrannen commented:
> However there is an issue pending that I'm not sure why but for some reason bnd is not identifying the import versions for project dependencies. This is strange since I do see that bnd has the correct classpath and the jars are there. For example `junit-platform-engine` does not properly import `org.junit.platform.commons.*` with versions. I need to consult with @bjhargrave to figure out what might be the cause.
> 
> Also pending is verification of the metadata.

OK. Please keep us posted on both accounts.

sbrannen commented on build.gradle.kts:
```diff
@@ -140,6 +142,31 @@ subprojects {
+
+						# The following exports only packages having classes annotated @API
+						# First collect all classes annotated with @API
+						possibleAPI: ${'$'}{classes;HIERARCHY_INDIRECTLY_ANNOTATED;org.apiguardian.api.API}
```
Cool. I didn't know you could do things like that with `bnd`. 👍 

But... that will automatically export `org.junit.jupiter.engine.config` (which should not be exported because all classes annotated with `@API` in that package are actually `@API(status = INTERNAL, ...)`) -- right?

So it looks like we will still have to fine tune things a bit.

Or can you do even more magic with `bnd` to check the `status` attribute of `@API` declarations?

rotty3000 commented:
Let's see what we can come up with. ☺️ It might take me a bit. If you like
I can mark the PR as WIP. Or maybe you have a label...

On Sat, Nov 9, 2019, 14:16 Sam Brannen, <notifications@github.com> wrote:

> *@sbrannen* commented on this pull request.
> ------------------------------
>
> In build.gradle.kts
> <https://github.com/junit-team/junit5/pull/2095#discussion_r344458083>:
>
> > @@ -140,6 +142,31 @@ subprojects {
>
>  		}
>
>
>
>  		afterEvaluate {
>
> +			if (project != project(":junit-jupiter")) {
>
> +				tasks["jar"].withConvention(aQute.bnd.gradle.BundleTaskConvention::class) {
>
> +					setBnd("""
>
> +						Bundle-Name: ${project.description}
>
> +						Import-Package: \
>
> +							!org.apiguardian.api,\
>
> +							kotlin.*;resolution:="optional",\
>
> +							*
>
> +						-fixupmessages.kotlin.import: "Unused Import-Package instructions: \\[kotlin.*\\]";is:=ignore
>
> +						-removeheaders: Private-Package
>
> +						-noextraheaders: true
>
> +
>
> +						# The following exports only packages having classes annotated @API
>
> +						# First collect all classes annotated with @API
>
> +						possibleAPI: ${'$'}{classes;HIERARCHY_INDIRECTLY_ANNOTATED;org.apiguardian.api.API}
>
>
> Cool. I didn't know you could do things like that with bnd. 👍
>
> But... that will automatically export org.junit.jupiter.engine.config
> (which should not be exported because all classes annotated with @API in
> that package are actually @API(status = INTERNAL, ...)) -- right?
>
> So it looks like we will still have to fine tune things a bit.
>
> Or can you do even more magic with bnd to check the status attribute of
> @API declarations?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/junit-team/junit5/pull/2095?email_source=notifications&email_token=AABD2TAKMOE6AMFUJGKJYXLQS4EAHA5CNFSM4JK5I3UKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCLAAEQI#pullrequestreview-314573377>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABD2TA5QA5RX544EM4HNC3QS4EAHANCNFSM4JK5I3UA>
> .
>

marcphilipp commented on settings.gradle.kts:
```diff
@@ -1,4 +1,12 @@
+			}
+		}
+	}
+
```
Please add the version to the `plugins` block instead and delete `resolutionStrategy` again.

rotty3000 replied:
Due to an issue with the gradle plugin portal reported here https://github.com/gradle/plugin-portal-requests/issues/9 version 4.3.0 is not accessible and so we have to use the approach shown until the issue can be resolved. Sorry about that.

marcphilipp replied:
I see. Thanks for the explanation!

marcphilipp commented on settings.gradle.kts:
```diff
@@ -10,6 +18,12 @@ pluginManagement {
+
+	repositories {
+		gradlePluginPortal()
+		mavenCentral()
```
Why do we need `mavenCentral()`?

rotty3000 replied:
Also due to the issue with the gradle plugin portal...

marcphilipp commented on build.gradle.kts:
```diff
@@ -140,6 +142,31 @@ subprojects {
 		afterEvaluate {
+			if (project != project(":junit-jupiter")) {
+				tasks["jar"].withConvention(aQute.bnd.gradle.BundleTaskConvention::class) {
+					setBnd("""
```
Doesn't `bnd = """` work?

rotty3000 replied:
```
Script compilation error:

  Line 147:                                     bnd = """
                 ^ Val cannot be reassigned
```

marcphilipp commented on build.gradle.kts:
```diff
@@ -140,6 +142,31 @@ subprojects {
 
 		afterEvaluate {
+			if (project != project(":junit-jupiter")) {
+				tasks["jar"].withConvention(aQute.bnd.gradle.BundleTaskConvention::class) {
```
Will this work for subprojects that have disabled the `jar` task and use `shadowJar` instead?

rotty3000 replied:
That's a good question. I only briefly looked at `shadowJar` but I wasn't clear it's real purpose. Now that you mention it I'll have to dig deeper into it. If you could fill me in with a small description of what the intention is there it'll help me make the best choice on how to handle it.

marcphilipp replied:
We use the shadow plugin to build a fat jar with relocated dependencies, e.g. in `junit-jupiter-params`. You could probably use the (not very pretty) approach from here to catch both tasks: https://github.com/junit-team/junit5/blob/master/buildSrc/src/main/kotlin/java-library-conventions.gradle.kts#L118

rotty3000 replied:
k, this is solved.

marcphilipp commented on build.gradle.kts:
```diff
@@ -140,6 +142,31 @@ subprojects {
 		}
 
 		afterEvaluate {
+			if (project != project(":junit-jupiter")) {
```
Why is this in an `afterEvaluate` block?

rotty3000 replied:
Because that was the only place I could get  `${project.description}` to resolve correctly. Outside of this I was getting null. Any suggestions to work around that would be appreciated.

rotty3000 replied:
this was also solved.

marcphilipp commented on build.gradle.kts:
```diff
@@ -140,6 +142,31 @@ subprojects {
 		}
 
 		afterEvaluate {
+			if (project != project(":junit-jupiter")) {
```
Please add this to `java-library-conventions.gradle.kts` instead of to the root build script.

rotty3000 replied:
Great, I wasn't aware of it's existence.

rotty3000 replied:
I managed to move the bnd configuration over to `java-library-conventions.gradle.kts`

marcphilipp commented on buildSrc/src/main/kotlin/java-library-conventions.gradle.kts:
```diff
@@ -180,6 +183,29 @@ tasks.jar {
+
+	withConvention(BundleTaskConvention::class) {
+		setBnd("""
+			Bundle-Name: ${project.description}
```
Isn't this evaluated before the descriptions are set?

rotty3000 replied:
I split this up so that only that one pieces is during `afterEvaluate`

marcphilipp commented on buildSrc/src/main/kotlin/java-library-conventions.gradle.kts:
```diff
@@ -180,6 +183,29 @@ tasks.jar {
+			# First collect all classes annotated with @API
+			possibleAPI: ${'$'}{classes;HIERARCHY_INDIRECTLY_ANNOTATED;org.apiguardian.api.API}
+			# Strip off the classNames to obtain a list of package names
+			possiblePackages: ${'$'}{replace;${'$'}{possibleAPI};\.[^.]*${'$'};}
```
I'm curious if we could we parse the module descriptors instead... 🤔

sbrannen replied:
That would be great it we could!

Better to not duplicate this information.

sormuras replied:
We could parse their source code (similar to [this](https://github.com/sormuras/bach/blob/master/src/bach/Bach.java#L697-L728)) or use the already compiled descriptors -- they are compiled at this stage, aren't they?

rotty3000 replied:
The module descriptor will always trail your actual code. You should let tooling give you a result _as close to perfect as a algorithm can bring it_ and then let you take it the last mile. This will eliminate guess work and automatically find details humans aren't good at identifying.

In short, let's leave that part for a future update, shall we? :)

rotty3000 replied:
This is not to say that I will not solve the @API exports issue... but rather then parsing of the module descriptor suggestion. ;)

rotty3000 replied:
... if that's ok with you all!

marcphilipp replied:
Fair enough, we could instead add an integration test that verifies the two things are aligned.

rotty3000 replied:
resolving this since we have a solution now

kriegfrj commented:
I like where this project is going, thanks guys. A couple of comments:

* Given the way that JUnit 5 is versioned, perhaps we could consider making use of OSGi contracts instead of straight-out import/export package mechanism? I don't have a lot of first-hand experience with OSGi contracts, but from what little knowledge I do have, they might be a fit. Perhaps @bjhargrave can comment on this.

* I would also like to see us leverage the OSGi capability mechanism in some way to define dependency relationships between test bundles and test engine bundles. That way, the resolver can help with making sure that the correct engines are included in the set of run bundles. For example, on the test engine bundle side, something like this:

```
Provide-Capability: org.junit.platform.engine;
    org.junit.platform.engine.id='junit-jupiter;version="[5.5.1]"' # By convention, make this == TestEngine.getId()
    version=5.5.1
```

...then on the test bundle side, you can declare (eg) the following:

```
Require-Capability: org.junit.platform.engine;
   org.junit.platform.engine.id='junit-jupiter;version="[5.5,6)",junit-vintage;version=4.12,org.myorg.my-test-engine;version=1.0'
```

Now the resolver will be able to ensure that the test bundles are paired with the correct versions of the correct test engines. The syntax might need a bit of tweaking (I'm not an expert in this area), but hopefully the idea makes sense.

rotty3000 commented:
Ok, only pending are:
- the `@API(status = INTERNAL, ...)`
- the verification

rotty3000 commented:
@kriegfrj let's not shock our goodly junit developers with too much too early :D 

However, I do like your suggestion. Maybe you could open a separate issue specific to capabilities and requirements?

kriegfrj commented:
> @kriegfrj let's not shock our goodly junit developers with too much too early :D
>

:smile: sorry!

> However, I do like your suggestion. Maybe you could open a separate issue specific to capabilities and requirements?

Sure - good idea to keep it separate. Where should I raise the issue - here, on bnd github, or both?

rotty3000 commented:
I'd suggest on this project since that's where you intended to make the change, IIUC.

rotty3000 commented:
Ok, so, the `@API(status = INTERNAL)` is solved. That also allowed me to do a big cleanup. the changes are much more concise again. You'll note that I added an _inline_ bnd plugin benefiting from the kotling buildSrc that was already there. But do note that bnd is adding support for APIGuardian in the future release (but I didn't want to wait so I added the plugin here).

Only verification remains.

rotty3000 commented:
Ok, so before we move on, I wanted to point out that all this did find some oddities:

- `org.junit.platform.commons.logging` is not exported because all it's types are marked as `@API(status = INTERLAN)` however it's imported by almost all other modules.

So, what we're thinking is that _really_ `@API` should always be exported, but that the package exports should also always be marked with their `status` attribute. In the case of INTERNAL, it's really an API for junit itself, like a `exports ... to ...` in JPMS. We can do that also in OSGi. So I'm going to change the heuristic.

sbrannen commented:
> Ok, so, the `@API(status = INTERNAL)` is solved. 

Great

> But do note that bnd is adding support for APIGuardian in the future release

Oh wow. That's a pleasant surprise.

> Only verification remains.

👍

rotty3000 commented:
Ok, so this PR is finally feature complete, including verification.

marcphilipp commented on gradle.properties:
```diff
@@ -17,3 +17,8 @@ org.gradle.jvmargs=-Xmx1g -XX:MaxMetaspaceSize=512m -XX:+HeapDumpOnOutOfMemoryEr
+# the jar is created and used instead of using src/main/classes.
+#
+# See https://docs.gradle.org/current/userguide/java_library_plugin.html#sub:java_library_known_issues_windows_performance
+systemProp.org.gradle.java.compile-classpath-packaging=true
```
Why is this required?

rotty3000 replied:
Ok, let me see. 

It begins with this very good question by @sbrannen:
> Why is org.opentest4j the only import with version information?

This was also a surprise to me because bnd gradle plugin was normally always able to determine the version to import a package because it would scan the classpath (essentially `project.sourceSet.main.compile`) and observe the package versions it found within the __jars__ that are there.

Remember, the bnd gradle plugin only adds metadata to the __jar__.

However, this was not working on this project! Why not?

I observed that this project uses the `java-library` gradle plugin which is an observably uncommon scenario as most gradle projects use the `java` plugin.

The distinctive difference here is that the `java` gradle plugin having dependencies of type `project` (e.g. `compile project(":foo")`) will result the project __jars__ being in `project.sourceSet.main.compile`. As I indicated earlier, this allowed bnd to observe the OSGi metadata for those dependent projects.

Now, the `java-library` plugin does _not_ by default cause a project dependency's jar to be built, AND when looking at the ``project.sourceSet.main.compile` will only contain the project output directory. Bnd does not emit metadata into the output directory and therefore there is no metadata to observer.

Setting the flag above to `true` tells the `java-library` plugin to revert to building the jar and using this in `project.sourceSet.main.compile`.

rotty3000 replied:
Also, as I indicated in the comment, you can read more about this here https://docs.gradle.org/current/userguide/java_library_plugin.html#sub:java_library_known_issues_windows_performance

rotty3000 replied:
Or better https://docs.gradle.org/current/userguide/java_library_plugin.html#sec:java_library_classes_usage

rotty3000 commented:
@marcphilipp I will add some comments around all the tasks to explain their use. Standby! :)

rotty3000 commented:
@marcphilipp I added lots of comments, but please if you have more questions I can take those too.

kriegfrj commented on buildSrc/build.gradle.kts:
```diff
@@ -10,4 +10,5 @@ repositories {
 dependencies {
 	implementation(kotlin("gradle-plugin"))
 	implementation("de.marcphilipp.gradle:nexus-publish-plugin:0.4.0")
+	implementation("biz.aQute.bnd:biz.aQute.bnd.gradle:4.3.0")
```
Should we bump to 4.3.1 now that the maintenance release is publicly available?

marcphilipp commented on junit-platform-console-standalone/junit-platform-console-standalone.gradle.kts:
```diff
@@ -59,4 +74,10 @@ tasks {
+	// So in order to resolve this, it can only run on Java 9
+	verifyOSGiProperties {
+		property("-runee", "JavaSE-9")
+	}
```
This feels wrong to me. Do you have a failure message?

rotty3000 replied:
I will check to see exactly what classes are compiled with Java 9 bytecode (bnd uses the class' magic number to make this determination).

rotty3000 replied:
The class in question is `org.junit.platform.console.ConsoleLauncherToolProvider` which implements `java.util.spi.ToolProvider` which is `@since 9` :)

marcphilipp commented on buildSrc/src/main/kotlin/java-library-conventions.gradle.kts:
```diff
@@ -139,6 +143,105 @@ if (project in mavenizedProjects) {
+			-fixupmessages.kotlin.import: "Unused Import-Package instructions: \\[kotlin.*\\]";is:=ignore
+
+			# This tells bnd to ignore classes it files in `META-INF/versions/`
+			# because bnd doesn't yet support multi-version jars.
```
```suggestion
			# because bnd doesn't yet support multi-release jars.
```

rotty3000 replied:
fixing

rotty3000 commented:
As requested I added a few checks to the manifest test.

rotty3000 commented:
failures here do not seem to be related. :(

rotty3000 commented:
See the same failures on this other sanity check PR https://github.com/junit-team/junit5/pull/2101

sormuras commented:
`Manifest attribute Bundle-Version expected to be 5.6.0-SNAPSHOT, but is: 5.6.0.SNAPSHOT`

There's `.` written to the manifest instead of a `-`. Who generates the string value that's used for `Bundle-Version`?

rotty3000 commented:
Ah, I didn't see this error. Dot is required by OSGi version syntax. I will parse the version into the correct syntax for this check. Standby!
